node/full_node_test.go

```diff
@@ -265,59 +265,137 @@ func TestPendingBlocks(t *testing.T) {
+		}
+		app.On("ExtendVote", mock.Anything, mock.Anything).Return(invalidVoteExtension)
+
+		// Ensure all necessary methods are mocked
+		app.On("Commit", mock.Anything, mock.Anything).Return(&abci.ResponseCommit{}, nil)
+		app.On("PrepareProposal", mock.Anything, mock.Anything).Return(&abci.ResponsePrepareProposal{Txs: nil}, nil)
+		app.On("ProcessProposal", mock.Anything, mock.Anything).Return(&abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_ACCEPT}, nil)
+		app.On("FinalizeBlock", mock.Anything, mock.Anything).Return(finalizeBlockResponse)
+		require.NotNil(app)
+
+		// Start the node
+		require.NoError(node.Start())
+
+		// Wait for blocks to be produced until voteExtensionEnableHeight
+		for i := 1; i < voteExtensionEnableHeight; i++ {
+			select {
+			case <-ctx.Done():
+				require.Fail("timeout waiting for block height")
+			default:
+				time.Sleep(100 * time.Millisecond) // Avoid busy waiting
+			}
+		}
 
-	// Create & configure node with app. Get signing key for mock functions.
-	app := &mocks.Application{}
-	app.On("InitChain", mock.Anything, mock.Anything).Return(&abci.ResponseInitChain{}, nil)
-	node, signingKey := createAggregatorWithApp(ctx, app, voteExtensionEnableHeight, t)
-	require.NotNil(node)
-	require.NotNil(signingKey)
+		status, err := node.GetClient().Status(ctx)
+		require.NoError(err)
+		require.EqualValues(voteExtensionEnableHeight-1, status.SyncInfo.LatestBlockHeight, "Expected block height mismatch")
+
+		// Wait for extendVoteFailure to be called, indicating a retry attempt
+		<-extendVoteFailureChan
+
+		// Ensure that the ExtendVote method is called with the expected arguments
+		app.AssertCalled(t, "ExtendVote", mock.Anything, mock.Anything)
 
-	prepareProposalVoteExtChecker := func(_ context.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
-		if req.Height <= voteExtensionEnableHeight {
-			require.Empty(req.LocalLastCommit.Votes)
-		} else {
-			require.Len(req.LocalLastCommit.Votes, 1)
-			extendedCommit := req.LocalLastCommit.Votes[0]
-			require.NotNil(extendedCommit)
-			require.Equal(extendedCommit.BlockIdFlag, cmproto.BlockIDFlagCommit)
-			// during PrepareProposal at height h, vote extensions from previous block (h-1) is available
-			require.Equal([]byte(fmt.Sprintf(expectedExtension, req.Height-1)), extendedCommit.VoteExtension)
-			require.NotNil(extendedCommit.Validator)
-			require.NotNil(extendedCommit.Validator.Address)
-			require.NotEmpty(extendedCommit.ExtensionSignature)
-			ok, err := signingKey.GetPublic().Verify(extendedCommit.VoteExtension, extendedCommit.ExtensionSignature)
+		// Check the node's behavior after encountering the invalid vote extension error
+		// verify that the block height has not advanced
+		status, err = node.GetClient().Status(ctx)
+		require.NoError(err)
+		require.EqualValues(voteExtensionEnableHeight-1, status.SyncInfo.LatestBlockHeight, "Block height should not advance after vote extension failure")
```

Comment:
Move this all into the for loop you have. There isn't a reason to have this outside the for loop since you then repeat the check.